lib/redis.js

```diff
@@ -1,79 +1,135 @@
 'use strict';
 
-const redis = require('redis');
+const Redis = require('ioredis');
 const helpers = require('./helpers');
 
-function createClient(settings, createNew) {
-  let client;
-  if (isClient(settings)) {
-    // We assume it's a redis client from either node_redis or ioredis.
+// todo utilize ioredis [options.keyPrefix]
+// todo maybe utilize ioredis redis.defineCommand for lua scripts
+
+const clientIDs = new WeakMap();
+
+class AbortError extends Error {
+  constructor() {
+    super('Connection is closed.');
+    this.name = this.constructor.name;
+  }
+}
+
+async function createClient(
+  settings,
+  { createNew = false, getClientID = false } = {}
+) {
+  let client,
+    isNewClient = true;
+  const clientType = getClientType(settings);
+  if (clientType === 'NodeRedis') {
+    throw new TypeError(
+      "clients produced by the NodeRedis library ('redis' on npm) are no longer supported as of bee-queue@2"
+    );
+  }
+  if (clientType) {
+    // We've actually been passed an existing client instance
     client = settings;
 
     if (createNew) {
-      // Both node_redis and ioredis support the duplicate method, which creates
-      // a new redis client with the same configuration.
       client = client.duplicate();
     } else if (isReady(client)) {
       // If we were given a redis client, and we don't want to clone it (to
       // enable connection sharing between Queue instances), and it's already
       // ready, then just return it.
-      return Promise.resolve(client);
-    } // otherwise, we wait for the client to be in the ready state.
-  } else {
-    // node_redis mutates the options object we provide it, so we clone the
-    // settings first.
-    if (typeof settings === 'object') {
-      settings = Object.assign({}, settings);
+      return client;
+    } else {
+      isNewClient = false;
     }
+  } else {
+    client = Redis.createClient(settings);
```

Comment:
This is deprecated, but we override it in tests and can't replace the bare function as easily as we can the `createClient` export.